src/workerd/jsg/jsg.c++

```diff
@@ -327,6 +327,9 @@ kj::Maybe<JsObject> Lock::resolveModule(kj::StringPtr specifier) {
   auto spec = kj::Path::parse(specifier);
   auto& info = JSG_REQUIRE_NONNULL(
       moduleRegistry->resolve(*this, spec), Error, kj::str("No such module: ", specifier));
+  if (!!requireEsm) {
+    JSG_REQUIRE(info.maybeSynthetic == kj::none, TypeError, "Main module must be an ES module.");
+  }
```

Comment:
Nit: This would get rid of an extra branch

```suggestion
  JSG_REQUIRE(!!requireEsm || info.maybeSynthetic == kj::none, TypeError, "Main module must be an ES module.");
```